### PR TITLE
Add aws_secret module for managing secretsmanager on AWS

### PR DESCRIPTION
**Pull request by rrey**
##### SUMMARY

This is a rework of pull request #40093 as the author does not answer anymore on the PR thread.
I've rework the module but kept the lookup plugin and test unchanged (I've applied @s-hertel comments though).

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME

`modules/cloud/amazon/aws_secret.py`

##### ANSIBLE VERSION

```paste below
ansible 2.6.0
  config file = None
  configured module search path = [u'/Users/remirey/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/remirey/.venv/grafana/lib/python2.7/site-packages/ansible
  executable location = /Users/remirey/.venv/grafana/bin/ansible
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```

### CODE REVIEW

**Comment by willthames** [comment]
My usual approach with IAM roles for Ansible is to use a name that is consistent between test runs - that way we can ensure that the role exists once and we don't need to give CI the CreateRole permission. 

Rather than `test-secrets-manager-role-{{ resource_prefix }}`, just use `test-secrets-manager-role`

**Comment by gundalow** [comment]
Looks like permissions will need to be given so this test can run:
I'll ask @s-hertel to take a look

```
                          Traceback (most recent call last):
  File "/tmp/ansible_iam_role_payload_27dTZh/__main__.py", line 249, in create_or_update_role
    role = connection.create_role(**params)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 320, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 623, in _make_api_call
    raise error_class(parsed_response, operation_name)
ClientError: An error occurred (AccessDenied) when calling the CreateRole operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=92943.67 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::966509639900:role/test-secrets-manager-role-shippable-92943-67
```

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (s-hertel) [inline test/integration/targets/aws_secret/tasks/main.yaml:105]: Maybe a module bug - this task is making changes so the assert below fails.

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (s-hertel) [inline test/integration/targets/aws_secret/tasks/main.yaml]: The above task fails with "The aws_secret module requires a region and none was found in configuration, environment variables or module parameters". If you add region: "{{ aws_region }}" it will start failing with "Failed to describe secret: Unable to locate credentials". This is the same for the next two tasks too.

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (s-hertel) [inline test/integration/targets/aws_secret/tasks/main.yaml]: The ansible_testing directory in "{{ tmp.path }}" hasn't been created. You could create it or just use `dest: "{{ tmp.path }}/hello_world.zip"`. The other tasks using that path also need to be fixed.

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_secret.py]: Use camel_dict_to_snake_dict on result for consistency with other AWS modules

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_secret.py:368]: Use camel_dict_to_snake_dict on result for consistency with other AWS modules. It looks like this is already happening for state == 'absent'

File: lib/ansible/plugins/lookup/aws_secret.py
Comment (s-hertel) [inline lib/ansible/plugins/lookup/aws_secret.py:82]: If this fails, it would be good to be able to fall back to instance role credentials if they exist. Here's a commit fixing that for another plugin for reference if it helps https://github.com/ansible/ansible/commit/d5a5e37fcf39de757ebe66bb239d9b9a5837b9ab#diff-fde53b7bca9c303ec46c0a71219bab47

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_secret.py:20]: Do you mind making this a list and adding @slapula (`Aaron Smith <ajsmith10381@gmail.com>`) as an additional author to this module and the plugin since this is based on his work? If he resurfaces he'll be able to get pinged on things to help provide `shipit`s and maintain the module and plugin.

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_secret.py:89]: It will help out people a lot if you can document the detailed contents of the return. You can do this pretty easily by running the module (you could run the integration tests), saving the returned json into a file, and then running https://github.com/ansible/ansible/blob/devel/hacking/return_skeleton_generator.py with the file.

**Review by s-hertel — CHANGES_REQUESTED** [review]
Thanks for picking this up to help get it over the line @rrey!

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (rrey) [reply test/integration/targets/aws_secret/tasks/main.yaml:105]: Noob question:
What is the ansible-test command that allow me to run the integration test just for aws_secret ?

I'm a bit cheating locally and I guess it will be a source of differences in the result...

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (rrey) [reply test/integration/targets/aws_secret/tasks/main.yaml]: indeed, done.

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (rrey) [reply test/integration/targets/aws_secret/tasks/main.yaml]: that' clearly one of the diff from my env. I'll reproduce this and check how to fix with the ansible-test command you'll provide :)

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py]: ok done

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py:368]: indeed, done.

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py:20]: No problem,  I changed it initially because I modified most of the code from the module.

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py:89]: Thanks for the tips ! awesome !

File: lib/ansible/plugins/lookup/aws_secret.py
Comment (rrey) [reply lib/ansible/plugins/lookup/aws_secret.py:82]: ok, I pushed a fix for your comments except this one.

This part of the code looks to be a copy of aws_ssm plugin. shouldn't the credentials retrieval be factorized somewhere in aws common code ?

**Comment by rrey** [comment]
Thanks for the review @s-hertel !

**Comment by slapula** [comment]
@rrey Sorry I've been out of pocket these past couple months.  I have cycles now to look at my Ansible PRs (including the secrets_manager module but I'll defer to this one since this work is more recent).  If you need any assistance with this please let me know.

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (mattclay) [reply test/integration/targets/aws_secret/tasks/main.yaml:105]: `ansible-test integration aws_secret -v`

**Comment by rrey** [comment]
@slapula : Thanks !
@s-hertel : Can you fix the issue with the role creation ?

```
 ClientError: An error occurred (AccessDenied) when calling the CreateRole operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=94536.67 is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::966509639900:role/test-secrets-manager-role
```

Should I drop the deletion of the role at the end of the tests to allow you to create it once and for all as suggested by @willthames  ?

**Comment by s-hertel** [comment]
Hi @rrey, I keep missing you on IRC until after you log off. There is a PR open to the CI repo for role creation (as well as the other permissions the tests need). I need a review on that pull request and I have requested @mattclay's decision on https://github.com/ansible/ansible/pull/48486#issuecomment-437713758 since we have tests doing different things and deciding on a common approach would be good. Once tests are running I'll be able to do a last review and merge. Thanks for your work on this!

**Comment by mente** [comment]
Follow up comment to @rrey  - as far as I see this module will not return the secret if needed. Will you be so kind to add this functionality as well?

I would love to contribute to this PR but I'm not a Python Developer, nor have experience with Ansible development (which is a bad excuse for open source project of course)

**Comment by rrey** [comment]
@s-hertel : Awesome ! Thx for the  feedback !

@mente : the modules aims at reaching  a state.
Either present and you provided the value.
Either absent and you don’t care about the value.

It is true that the module could return the value in its output, but the idempotency implies that you provided it to make the diff. So returning it won’t provide anything new in your context.

If you want to retrieve the value, you should use the lookup plugin which is part of this PR as well (aws_secret).

**Comment by mattclay** [comment]
There are basically two options for creating the role:

- Dynamic (`test-secrets-manager-role-{{ resource_prefix }}`) - Automatically created and removed during the test run.
  - Pro: Easy to support in CI (allow limited CreateRole and attach AttachRolePolicy, DetachRolePolicy).
  - Con: Requires CreateRole, AttachRolePolicy, DetachRolePolicy permissions (cannot create role in advance).
- Static (`test-secrets-manager-role`) - Created manually before the test run or automatically during the test run if not already present. Not automatically removed.
  - Pro: Does not require CreateRole, AttachRolePolicy, DetachRolePolicy permissions (can create role in advance).
  - Con: More complex to support in CI (allow limited CreateRole and attach AttachRolePolicy, exclude role from cleanup process).

Although the static option is more work to implement in CI it should provide a better user experience for anyone trying to run the tests using their own account. @s-hertel Do you agree?

**Comment by s-hertel** [comment]
Yeah, I think the static option has the best of both worlds. Thanks for laying out the options @mattclay. 

@rrey since it looks like you've removed the resource_prefix from the role name, all that is left to do is remove the task that deletes the IAM role.

**Comment by rrey** [comment]
done ! :) @s-hertel : let me know when the CI PR is merged.

The only remaining open point is [on this comment](https://github.com/ansible/ansible/pull/48486#discussion_r233109660).

What's your guideline ? Should I apply the code portion that you pointed in this PR ?
IMHO the credentials retrieval should be managed somewhere in the AWS helpers and the lookups, modules (etc ...) should be updated to use that common code to have everything doing the same thing.

**Comment by s-hertel** [comment]
@rrey Modules do use common code for credentials but not plugins yet, as there have only been 2 with a consistent approach. There are some inconsistencies with the existing lookup plugins and credentials. My initial thought was that I'd like to see those problems fixed across the board and there's already a lot going on in this PR. But feel free to do the initial step of moving the common code for this PR, aws_ec2, and aws_rds into module_utils.

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (s-hertel) [inline test/integration/targets/aws_secret/tasks/main.yaml]: For tests to pass this needs to be 
```suggestion
      rotation_lambda: "arn:aws:lambda:{{ aws_region }}:{{ test_caller_facts.account }}:function:hello-world-{{ resource_prefix }}"
```

**Comment by rrey** [comment]
>  But feel free to do the initial step of moving the common code for this PR, aws_ec2, and aws_rds into module_utils.

@s-hertel : I'll check this after this PR is merged, but why not !

**Comment by s-hertel** [comment]
@rrey I haven't forgotten your PR. I have tried to deploy the PR for this and we've hit the maximum length of the CI policy. To be able to merge it I have refactored the CI policies it in a large PR with @mattclay's guidance. Then the PR for CI for this PR can be deployed and merged. You don't need to take further action for this PR. I'm on PTO today and tomorrow. Hopefully, by sometime next week my large PR for CI will be able to receive a thorough review.

**Comment by rrey** [comment]
@s-hertel : thanks a lot for your availability on this PR !

**Comment by s-hertel** [comment]
Received initial feedback on the massive PR and addressed it, still needs a thorough review.

**Comment by rrey** [comment]
@s-hertel any news ?

File: test/integration/targets/aws_secret/tasks/main.yaml
Comment (s-hertel) [reply test/integration/targets/aws_secret/tasks/main.yaml:105]: @rrey This one is still making changes to the secret so the tests fail. Looks like an idempotence issue with the module. My PR restructuring things for CI has been merged and once your tests are completing rather than failing here the PR adding the permissions and this can be merged. Here's the output from testing on the dev server:
```
TASK [aws_secret : add secret to AWS Secrets Manager] **************************
changed: [testhost] => {"changed": true, "secret": {"arn": "arn:aws:secretsmanager:us-east-1:966509639900:secret:test-secret-string-ansible-test-macbook-pro-44-50736924-4CXvRp", "last_changed_date": "2019-02-21T19:44:25.772000+00:00", "name": "test-secret-string-ansible-test-shertel-50736924", "version_ids_to_stages": {"26d43df3-84a1-43eb-803e-40f0f0c65d1d": ["AWSCURRENT"]}}}

TASK [aws_secret : assert correct keys are returned] ***************************
ok: [testhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

TASK [aws_secret : no changes to secret] ***************************************
changed: [testhost] => {"changed": true, "secret": {"arn": "arn:aws:secretsmanager:us-east-1:966509639900:secret:test-secret-string-ansible-test-shertel-50736924-4CXvRp", "last_accessed_date": "2019-02-21T00:00:00+00:00", "last_changed_date": "2019-02-21T19:44:26.995000+00:00", "name": "test-secret-string-ansible-test-shertel-50736924", "version_ids_to_stages": {"26d43df3-84a1-43eb-803e-40f0f0c65d1d": ["AWSPREVIOUS"], "6ad8339b-27a9-4606-938e-8d2f0c5979eb": ["AWSCURRENT"]}}}

TASK [aws_secret : assert correct keys are returned] ***************************
fatal: [testhost]: FAILED! => {
    "assertion": "not result.changed",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by rrey** [comment]
Hi @s-hertel, I fail to reproduce any failure with my own aws account and the integration tests.
I forced push to force the CI to re-run, but the error I see in the 2.7 execution is still related to permissions and IAM. :(

**Comment by s-hertel** [comment]
@rrey Okay, thanks for trying to take a look. The tests are still failing there for me so I'm working on debugging/fixing them since I have the reproducer. Once the tests are passing for me I can deploy to production and rerun shippable on this PR (no rebasing necessary on your part).

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/aws_secret.py]: The failures were because of this line. The value was always changing on Python 3 because the current value was encoded as a byte string whereas the desired secret is unicode.

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py]: Thanks for your time on this. I should have been able to reproduce this :(

File: test/integration/targets/aws_secret/aliases
Comment (s-hertel) [inline test/integration/targets/aws_secret/aliases:2]: Mattclay pointed out this morning that there are some issues adequately resource restricting creating secrets. I'm going to look at addressing that after freeze for 2.8 and I don't want this module to miss the deadline so I've disabled the tests (which are passing locally for me with Python 2 and 3).

File: lib/ansible/modules/cloud/amazon/aws_secret.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/aws_secret.py]: @rrey It's no problem. Do my commits look okay? If so, once shippable completes I'll hit merge.

Comment (rrey) [reply lib/ansible/modules/cloud/amazon/aws_secret.py]: lgtm :)

File: test/integration/targets/aws_secret/aliases
Comment (mattclay) [inline test/integration/targets/aws_secret/aliases]: This alias should be removed.

**Comment by s-hertel** [comment]
Thanks @rrey for the time you've spent on this!

**Comment by mnylen** [comment]
The documentation is a bit misleading at https://docs.ansible.com/ansible/devel/plugins/lookup/aws_secret.html - it says this is new in 2.7, but as far as I know, this lookup plugin hasn't been released yet in any ansible version? @s-hertel 

I spent a few hours trying to figure out why my 2.7.8 ansible was complaining that aws_secret lookup plugin isn't available. :)

**Comment by rrey** [comment]
@mnylen you are right it was originally targeting 2.7, but it missed the milestone.
Here is the patch : https://github.com/ansible/ansible/pull/53390

**Pull request merged**
